### ISSUE TITLE
Changes on #9825 caused an issue

### ISSUE THREAD

**Issue by hikalkan** [dotnet/efcore#10407]
Hi all,

#### Problem

The changes to solve this issue caused an inconsistent exception for us. We disabled this feature by the code below:

````C#
AppContext.SetSwitch("Microsoft.EntityFrameworkCore.Issue9825", true);
````

Then our code worked again. The exception we have (without enabling Microsoft.EntityFrameworkCore.Issue9825 as shown above):

````C#
System.InvalidOperationException: variable 'context' of type 'Abp.EntityFrameworkCore.AbpDbContext' referenced from scope '', but it is not defined
   at System.Linq.Expressions.Compiler.VariableBinder.Reference(ParameterExpression node, VariableStorageKind storage)
   at System.Linq.Expressions.Compiler.VariableBinder.VisitParameter(ParameterExpression node)
   at System.Linq.Expressions.ParameterExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitMember(MemberExpression node)
   at System.Linq.Expressions.MemberExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
   at System.Linq.Expressions.Compiler.VariableBinder.VisitInvocation(InvocationExpression node)
   at System.Linq.Expressions.InvocationExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Linq.Expressions.Compiler.VariableBinder.VisitUnary(UnaryExpression node)
   at System.Linq.Expressions.UnaryExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
   at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
   at System.Linq.Expressions.Compiler.VariableBinder.VisitLambda[T](Expression`1 node)
   at System.Linq.Expressions.Expression`1.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   at System.Linq.Expressions.Compiler.LambdaCompiler.Compile(LambdaExpression lambda)
   at System.Linq.Expressions.Expression`1.Compile(Boolean preferInterpretation)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateExecutorLambda[TResults]()
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateAsyncQueryExecutor[TResult](QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileAsyncQuery[TResult](QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileAsyncQueryCore[TResult](Expression query, INodeTypeProvider nodeTypeProvider, IDatabase database)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass24_0`1.<CompileAsyncQuery>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddAsyncQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileAsyncQuery[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ExecuteAsync[TSource,TResult](MethodInfo operatorMethodInfo, IQueryable`1 source, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.CountAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
   at Volo.FootbalManager2.Authorization.Users.UserLinkAppService.<GetLinkedUsers>d__7.MoveNext() in C:\Users\ismcagdas\Downloads\FootbalManager2\aspnet-core\src\Volo.FootbalManager2.Application\Authorization\Users\UserLinkAppService.cs:line 76
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at lambda_method(Closure , Object )
   at Microsoft.Extensions.Internal.ObjectMethodExecutorAwaitable.Awaiter.GetResult()
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeActionMethodAsync>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeNextActionFilterAsync>d__10.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow(ActionExecutedContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeInnerFilterAsync>d__14.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeNextExceptionFilterAsync>d__23.MoveNext()
````

#### Query & QueryFilter:

Our query is like that:

````C#
private IQueryable<LinkedUserDto> CreateLinkedUsersQuery(UserAccount currentUserAccount, string sorting)
{
    var currentUserIdentifier = AbpSession.ToUserIdentifier();

    return (from userAccount in _userAccountRepository.GetAll()
            join tenant in _tenantRepository.GetAll() on userAccount.TenantId equals tenant.Id into tenantJoined
            from tenant in tenantJoined.DefaultIfEmpty()
            where
                (userAccount.TenantId != currentUserIdentifier.TenantId ||
                userAccount.UserId != currentUserIdentifier.UserId) &&
                userAccount.UserLinkId.HasValue &&
                userAccount.UserLinkId == currentUserAccount.UserLinkId
            select new LinkedUserDto
            {
                Id = userAccount.UserId,
                TenantId = userAccount.TenantId,
                TenancyName = tenant == null ? "." : tenant.TenancyName,
                Username = userAccount.UserName,
                LastLoginTime = userAccount.LastLoginTime
            }).OrderBy(sorting);
}
````

And Global Query filter is like that:

https://github.com/aspnetboilerplate/aspnetboilerplate/blob/dev/src/Abp.EntityFrameworkCore/EntityFrameworkCore/AbpDbContext.cs#L153-L193

(if the return value is null, we don't call ``HasQueryFilter``, so it's not a problem. I also tried to remove CombineExpressions and combine them manually in a single expression).

#### Extra Info

Unfortunately, this issue is very inconsistent.

* Could not repeat it with integration tests (used SQLite in-memory db).
* This issue only a problem when our query is first executed. It does not throw exception for subsequent calls.

**Comment by ajcvickers** [comment]
Possible duplicate of #10301

**Comment by maliming** [comment]
waiting...

**Comment by smitpatel** [comment]
@hikalkan - We would need a simplified repro for this. Based on the query filter definition, you are configuring filters inside your DbContext class only so it should work with 2.0.1 package. Looking at error since parameter is not found in scope, perhaps the way expressions are combined (and/or created) is the thing which is throwing expression tree off.

**Comment by hikalkan** [comment]
Combining expressions is also inside the code: https://github.com/aspnetboilerplate/aspnetboilerplate/blob/dev/src/Abp.EntityFrameworkCore/EntityFrameworkCore/AbpDbContext.cs#L478

However, it's hard to reproduce it in a simple code. I was hoping that you can somehow guess it :(

**Comment by smitpatel** [comment]
@hikalkan - That was my guess that Expression tree became inconsistent else it wouldn't throw incorrect scope error. Perhaps before the patch change it was getting ignored rather than parametrized and now this change is causing to inject parameter but incorrect scope.

Fix for #10301 simplifies and make the pipeline more robust (and has a lot of patterns for testing). Would you be able to try out nightly build after that fix is in?

That could save both of us trouble of finding a repro. If not, I will give it another shot if I can extract out parts and reproduce the error.

**Comment by smitpatel** [comment]
@hikalkan - I am not able to extract out a repro code for this. Using interfaces and CombineExpressions in that way with a group join query works fine for me in 2.0.1 release and 2.1.0-preview1 packages. To process this any further we'll need a repro code which shows above behaviour.
